### PR TITLE
WIP - Design Proposal for functionVersioning

### PR DESCRIPTION
**Pull request by smruthi2187**
This is just a proposal to support function versioning in fission. Please suggest any modifications or feel free to propose a different approach towards function versioning.
Thanks.

---
This change is [Reviewable](https://reviewable.io/reviews/fission/fission/438)

### CODE REVIEW

File: Documentation/wip/FunctionVersioningProposal.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+	FunctionRef FunctionRef
+} 
+```
+* Introducing a new CRD of type `"FunctionMaster"`. This will have a reference to all aliases of a function. Also will have a list of references to all versions created for this function.
+```
+Type FunctionMaster struct {
+	Name string
+	Alias []Alias
+	Versions []FunctionRef
+}
```
Comment (vishal-biyani) [inline Documentation/wip/FunctionVersioningProposal.md]: How is an Alias mapped to a certain version of FunctionRef, just thinking if a Map might better suit here.

File: Documentation/wip/FunctionVersioningProposal.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+Type Tag struct {
+	Name string
+	FunctionRef []FunctionRef
+}
+```
+
+#### General Flow 
+1. When a user wants to create a function with different versions, he can do so with the following commands
+    ```bash
+    $ fission function version create --name hello --env nodejs --code hello1.js --version v1.0.0
```
Comment (vishal-biyani) [inline Documentation/wip/FunctionVersioningProposal.md]: This is my personal opinion but using a SemVar like versioning (i.e. x.y.z) makes it a bit user un-friendly and might be an overkill. I like simpler naming scheme with a single number. Also in practice how many versions of a function might user create? Even if it is in range of 1000, the single number versioning is far simpler. Open to more discussion on this.

File: Documentation/wip/FunctionVersioningProposal.md (hunk 1 above)
Comment (smruthi2187) [reply Documentation/wip/FunctionVersioningProposal.md]: For each version of the function, this design intends to create a new function object, with name "functionName_versionValue". The Alias type holds a reference to a specific functionName. Actually, the "General Flow" explains the flow of commands that the user would use to create different versions of the function.

File: Documentation/wip/FunctionVersioningProposal.md (hunk 2 above)
Comment (smruthi2187) [reply Documentation/wip/FunctionVersioningProposal.md]: Yes, I agree too. But it's upto the user to create versions according to any syntax. I've also listed this point under "Note" section in the end. I'm not very keen on fission maintaining the version values.

**Comment by smruthi2187** [comment]
Not valid anymore.

**Pull request closed without merging**